src/auth/src/credentials/mds_credential.rs

```diff
@@ -13,37 +13,164 @@
 // limitations under the License.
 
 use crate::credentials::dynamic::CredentialTrait;
-use crate::credentials::{Credential, Result};
+use crate::credentials::{Credential, Result, DEFAULT_UNIVERSE_DOMAIN, QUOTA_PROJECT_KEY};
 use crate::errors::{is_retryable, CredentialError};
 use crate::token::{Token, TokenProvider};
 use async_trait::async_trait;
 use bon::Builder;
 use http::header::{HeaderName, HeaderValue, AUTHORIZATION};
 use reqwest::Client;
+use reqwest::StatusCode;
+use std::default::Default;
 use std::sync::Arc;
 use std::time::Duration;
+use tokio::sync::watch;
+use tokio::sync::Notify;
 
 const METADATA_FLAVOR_VALUE: &str = "Google";
 const METADATA_FLAVOR: &str = "metadata-flavor";
 const METADATA_ROOT: &str = "http://metadata.google.internal/computeMetadata/v1";
 
 pub(crate) fn new() -> Credential {
-    let token_provider = MDSAccessTokenProvider::builder()
-        .endpoint(METADATA_ROOT)
-        .build();
-    Credential {
-        inner: Arc::new(MDSCredential { token_provider }),
-    }
+    MDSCredentialBuilder::default().build()
+}
+
+#[derive(Debug)]
+struct UniverseDomainCache {
+    universe_domain: Option<String>,
+    is_cached: bool,
 }
 
 #[derive(Debug)]
 struct MDSCredential<T>
 where
     T: TokenProvider,
 {
+    quota_project_id: Option<String>,
+    ud_rx: watch::Receiver<UniverseDomainCache>,
+    wakeup_tx: Arc<Notify>,
     token_provider: T,
 }
 
+#[derive(Debug, Default)]
+pub struct MDSCredentialBuilder {
+    endpoint: Option<String>,
+    quota_project_id: Option<String>,
+    scopes: Option<Vec<String>>,
+    universe_domain: Option<String>,
+}
+
+#[allow(dead_code)]
+impl MDSCredentialBuilder {
+    pub fn endpoint<S: Into<String>>(mut self, endpoint: S) -> Self {
+        self.endpoint = Some(endpoint.into());
+        self
+    }
+
+    pub fn quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    pub fn universe_domain<S: Into<String>>(mut self, universe_domain: S) -> Self {
+        self.universe_domain = Some(universe_domain.into());
+        self
+    }
+
+    pub fn scopes<S: Into<String>>(mut self, scopes: Vec<S>) -> Self {
+        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
+        self
+    }
+
+    pub fn build(&self) -> Credential {
```

Comment:
I think you want:

```suggestion
    pub fn build(self) -> Credential {
```

You can save some cloning I presume.